python/ray/train/trainer.py

```diff
@@ -330,10 +337,20 @@ def run(
                 checkpoint_strategy=checkpoint_strategy,
                 run_dir=self.latest_run_dir,
             )
+            aggregated_results = collections.defaultdict(list)
+            if aggregate_funcs is None or len(aggregate_funcs) == 0:
+                aggregate_funcs = {}
+            elif isinstance(aggregate_funcs, list):
+                aggregate_funcs = {e.__name__: e for e in aggregate_funcs}
+
             for intermediate_result in iterator:
+                for aggregate_name, func in aggregate_funcs.items():
+                    aggregated_results[aggregate_name].append(func(intermediate_result))
```

Comment:
While this does aggregate metrics, I think right now it may not be able to address one of the core usability aspects, which is to be able to track the metrics properly via Callbacks (e.g. `TBXLoggerCallback` for TensorBoard).

I've cc'ed @Yard1 who has done some [cool work](https://github.com/Yard1/ray-skorch/blob/main/ray_skorch/callbacks/train.py) in a neighboring library and may be able to share some advice/thoughts on how to design this so it's able to mesh well with Callbacks.

Reply:
@matthewdeng Thank you for the feedback! I didn't take callbacks into consideration. Will work more on this direction. Looking forward to hearing from suggestions from @Yard1.

Reply:
Yeah, we definitely would want to aggregate the metrics before they are passed to callbacks. I think it would make sense to have a magic key for aggregation in the worker dict, and add special cases for such keys wherever it's required.

Reply:
@Yard1 Thanks for your suggestions! I have a few questions. 
1. What are those magic keys? Did you mean some default aggregation methods like [these](https://github.com/Yard1/ray-skorch/blob/main/ray_skorch/callbacks/train.py#L26-L43)?
2. The aggregation can be done in [`__next__()`](https://github.com/ray-project/ray/blob/master/python/ray/train/trainer.py#L714) or [`_fetch_next_result()`](https://github.com/ray-project/ray/blob/master/python/ray/train/trainer.py#L730). Which one looks better? (I think `__next__()` can be a good choice.)
3. It seems you suggested separating aggregation from callbacks. How do we address @matthewdeng's comment on allowing metrics tracking at the same time?

Reply:
1. Magic keys in the `intermediate_results` passed to the callbacks - atm this is a list with each element corresponding to a worker. I think we could turn this into a dict/a list subclass and add an `aggregate` key there, so it can be accessed in a natural way from callbacks. That would be an API change, though, so we may have to think it through first. Another idea could be to pass the aggregate results in a separate argument to a callback.
2. I'd make a separate method and put it in `__next__`
3. What we want is to pass aggregate metrics to Callbacks - that way we can use those in our loggers.

Reply:
Sounds great! Thanks for your clarification. I will revise the implementation accordingly. Thanks 😊